Выполнить ФЗ 21

Volckoff opened the pull request:
Изменить Архитектуру БД. Создать свою схему для каждого микросервиса.

Реализовать модули для интернет магазина. Использовать FeignClient для общения между микросервисами. Реализовать логику конфигурации через ConfigServer и реализовать подключение через Eureka. Реализовать использование паттерна Circuit Breaker при обращении к сервису склада.

Volckoff committed:
Создать config-server

Volckoff committed:
Создать discovery-server

Volckoff committed:
Выполнить ФЗ 21.

Изменить Архитектуру БД. Создать свою схему для каждого микросервиса.

Реализовать модули для интернет магазина. Использовать FeignClient для общения между микросервисами. Реализовать логику кофигурации через ConfigServer и реализовать подключение через Eureka. Реализовать использование паттерна Circuit Breaker при обращении к сервису склада.

Volckoff committed:
Исправить pom для config и discovery

Volckoff commented:
Александр, добрый день!

Интересное задание, попытался реализовать все тонкости указанные в ТЗ (попробовал даже добавить FallBack клиент). Надеюсь с настройками не перемудрил.
В очередной раз столкнулся с проблемой при использовании MupStruct (ну нет у меня дружбы с этой библиотекой), не стабильно генерируются классы. Думаю что это моя ошибка, не до конца разобрался с взаимодействием зависимости MapStruct и build-helper-maven-plugin. Что - то с версиями напутал. Но разобраться не сумел. Классы генерировались но почему-то при попытке запуска приложения выдавал ошибку что не может найти сгенерированные классы. И так как все сроки уже сгорели, было принято волевое решение перейти на ручные маперы. Надеюсь это не ошибка.
Так же при локальном запуске все было хорошо, однако при запуске на GH выдал ошибку манифеста, забыл добавить spring-boot-maven-plugin и jar файлы не создавались.

avfyodorov commented:
> Александр, добрый день!
> 
> Интересное задание, попытался реализовать все тонкости указанные в ТЗ (попробовал даже добавить FallBack клиент). Надеюсь с настройками не перемудрил. В очередной раз столкнулся с проблемой при использовании MupStruct (ну нет у меня дружбы с этой библиотекой), не стабильно генерируются классы. Думаю что это моя ошибка, не до конца разобрался с взаимодействием зависимости MapStruct и build-helper-maven-plugin. Что - то с версиями напутал. Но разобраться не сумел. Классы генерировались но почему-то при попытке запуска приложения выдавал ошибку что не может найти сгенерированные классы. И так как все сроки уже сгорели, было принято волевое решение перейти на ручные маперы. Надеюсь это не ошибка. Так же при локальном запуске все было хорошо, однако при запуске на GH выдал ошибку манифеста, забыл добавить spring-boot-maven-plugin и jar файлы не создавались.

Конечно это не ошибка. Вы сейчас нарабатываете навыки, свой стиль программирования, поэтому самое важное это не использование каких-то библиотек, a чтобы лично Вам было удобно и понятно.

avfyodorov commented on commerce/shopping-cart/src/main/java/ru/practicum/ShoppingCartApp.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.cloud.openfeign.EnableFeignClients;
+
+@SpringBootApplication
+@EnableFeignClients
```
Необязательно, для информации.
В принципе, можно было бы указать какие именно клиенты стартовать и какие именно пакеты сканировать, у аннотации есть соответствующие параметры.
Таким образом будут инициализироваться только нужные клиенты. Лучше явно указать, что именно нужно, потому что в больших проектах могут быть и конфликты.

Volckoff replied:
Исправлено, добавлено точечное добавление инициализации клиента склада.
Я так понял что есть 2 возможности конфигурации:
1 - как я сделал, точечно указать какой клиент должен быть инициализирован
@EnableFeignClients(clients = {WarehouseClient.class})

2 - указать область сканирования для нахождения пакета
(basePackages = "ru.practicum.clients.warehouse")

Решил, что раз точно знаю какой клиент используется, то будет правильно его указать. А так по-моему разницы особо нет.

avfyodorov commented on commerce/shopping-store/src/main/java/ru/practicum/ShoppingStoreApp.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.cloud.openfeign.EnableFeignClients;
+
+@SpringBootApplication
+@EnableFeignClients
```
Аналогично... указать какие именно клиенты стартовать

Volckoff replied:
ShoppingStoreApp - не использует Feign‑клиенты.
Можно, в принципе совсем удалить @EnableFeignClients, однако пока решил оставить на случай если в следующем спринте будем расширять этот функционал.

avfyodorov replied:
вопрос снят )
Не заметил. (

avfyodorov commented on commerce/warehouse/src/main/java/ru/practicum/WarehouseApp.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.cloud.openfeign.EnableFeignClients;
+
+@SpringBootApplication
+@EnableFeignClients
```
И здесь какие именно клиенты стартовать

Volckoff replied:
Аналогично предыдущему, WarehoseApp - не использует Feign‑клиенты.
Можно, в принципе совсем удалить @EnableFeignClients, однако пока решил оставить на случай если в следующем спринте будем расширять этот функционал.

avfyodorov commented on commerce/shopping-store/src/main/java/ru/practicum/exceptions/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,39 @@
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@RestControllerAdvice
+public class GlobalExceptionHandler {
```
В класс хорошо бы добавить логирование.  Это полезно при разборе логов. Помогает быстрее понять когда и в каком случае возникла та или иная ошибка.

Volckoff replied:
Исправлено, добавлено логирование в обработчик для витрины. 
Почему - то не добавил ранее, наверное потому что не дебажил особо сильно этот класс.

avfyodorov commented on commerce/shopping-store/src/main/java/ru/practicum/repository/ProductRepository.java:
```diff
@@ -0,0 +1,16 @@
+
+import java.util.UUID;
+
+@Repository
```
Аннотация `@Repository` не требуется, так как наследуясь от JpaRepository - мы уже объявляем бин.

Volckoff replied:
Исправлено, аннотация удалена.

avfyodorov reviewed (commented):
Добрый день, Владимир Михайлович! 

Хорошая работа, всего несколько небольших уточнений.

Volckoff commented:
С MupStruct - просто обидно что я уже который спринт головой в эту библиотеку бьюсь, и все трудности какие-то. Все равно конечно ее освою, т.к. это - "стильно, модно, молодёжно". В общем широко распространено и применяется.
 Но пока вот так )

Volckoff committed:
Внести правки согласно комментариям ревью:
Добавить логирование в обработчик ошибок для витрины, Удалить ненужные аннотации, Добавить точечную инициализацию FeignClient.

avfyodorov reviewed (commented):
Добрый вечер, Владимир Михайлович!

Замечаний нет.
Работа принята.
